source/common/network/apple_dns_impl.cc

```diff
@@ -0,0 +1,259 @@
+    } catch (const std::exception& e) {
+      ENVOY_LOG(critical, "std::exception in DNSService callback: {}", e.what());
+      (*it)->dispatcher_.post([s = std::string(e.what())] { throw EnvoyException(s); });
+    } catch (...) {
+      ENVOY_LOG(critical, "Unknown exception in DNSService callback");
+      (*it)->dispatcher_.post([] { throw EnvoyException("unknown"); });
+    }
+
+    if ((*it)->owned_) {
+      ENVOY_LOG(warn, "Resolution for {} completed (async)", (*it)->dns_name_);
+      delete *it;
+    } else {
+      ENVOY_LOG(warn, "Resolution for {} completed (synchronously)", (*it)->dns_name_);
+      (*it)->synchronously_completed_ = true;
+    }
+  }
+
+  // Purge the contents so no one tries to delete them again.
+  queries_with_pending_cb_.clear();
+}
+
+AppleDnsResolverImpl::PendingResolution::~PendingResolution() {
+  ENVOY_LOG(warn, "Destroying PendingResolution for {}", dns_name_);
+  DNSServiceRefDeallocate(individual_sd_ref_);
+}
+
+void AppleDnsResolverImpl::PendingResolution::cancel() {
+  ENVOY_LOG(warn, "Cancelling PendingResolution for {}", dns_name_);
+  ASSERT(owned_);
+  if (pending_cb_) {
+    parent_.queries_with_pending_cb_.remove(this);
+  }
+  delete this;
+}
+
+void AppleDnsResolverImpl::PendingResolution::onDNSServiceGetAddrInfoReply(
+    DNSServiceFlags flags, uint32_t interface_index, DNSServiceErrorType error_code,
+    const char* hostname, const struct sockaddr* address, uint32_t ttl) {
+  ENVOY_LOG(warn,
+            "DNS for {} resolved with: flags={}[MoreComing={}, Add={}], interface_index={}, "
+            "error_code={}, hostname={}",
+            dns_name_, flags, flags & kDNSServiceFlagsMoreComing ? "yes" : "no",
+            flags & kDNSServiceFlagsAdd ? "yes" : "no", interface_index, error_code, hostname);
+  ASSERT(interface_index == 0);
+
+  if (error_code == kDNSServiceErr_DefunctConnection || error_code == kDNSServiceErr_Timeout ||
+      error_code == kDNSServiceErr_Refused) {
+    // Current query gets a failure status
+    if (!pending_cb_) {
+      ENVOY_LOG(warn, "[Error path] Adding to queries pending callback");
+      pending_cb_ = {ResolutionStatus::Failure, {}};
+      parent_.addPendingQuery(this);
+    } else {
+      ENVOY_LOG(warn, "[Error path] Changing status for query already pending flush");
+      pending_cb_->status_ = ResolutionStatus::Failure;
+    }
+
+    ENVOY_LOG(warn, "[Error path] DNS Resolver flushing queries pending callback");
+    parent_.flushPendingQueries();
+
+    // Deallocation of the MainSdRef __has__ to happend __after__ flushing queries. Flushing queries
+    // deallocates individual refs, so deallocating the main ref aheadd would cause deallocation of
+    // invalid individual refs per dns_sd.h
+    parent_.deallocateMainSdRef();
+    // POINT FOR DISCUSSION: can this throw here? Or does it need to be posted like the callback in
+    // flushPendingQueries.
+    parent_.initializeMainSdRef();
```

Comment:
This is super odd to me, but I don't know the API. Why are you removing/initializing the main API reference on each DNS query result? It seems like this would be done in the constructor resolver/destructor only?

Reply:
In reading more I guess this means the entire DNS connection is dead or something? For readability I would move all of this logic into the parent with better comments.

Reply:
> Why are you removing/initializing the main API reference on each DNS query result
Yep, your second comment got it right, this is __only__ happening on the error path. I can encapsulate and comment better.